app/models/comment.rb

```diff
@@ -64,6 +75,12 @@ def initialize(person, target, text)
       super(person, target)
     end
 
+    def build(*args)
+      super.tap do |comment|
+        comment.filter_mentions(comment.people_allowed_to_be_mentioned)
```

Comment:
> So what you built now is, that you filter all mentions on private posts on receive, which makes mention impossible on private posts.

Not really. Have a look on [this](https://github.com/diaspora/diaspora/pull/6818/files#diff-2bcec3a9d410c74617f05ba3d8408810R156) test -- it passes well, that means that mentions work on private receive. Actually, what I built now is no filtering at all (in the federation case). It rather creates mentions only for people who are allowed to be mentioned. However it doesn't work well right now in the case of multiple receive again.

The question is when do we create Mention objects for post/comment. Mention objects was originally created only when mention is allowed. That's why filtering was used - to avoid creation of Mention objects. If we change the logic of stream rendering so it hides posts when person is not allowed to see them irrespective of Mention object presence, then we could create Mention objects for all mentioned people immediately after receive. I guess this is what you suggest.

After some thoughts, I agree that this is more optimal way to implement visibility restriction for mentioning. This way we won't need any filtering at all and get more straightforward design.